File: MyAFBase/MyAFBase/Views/Home/HomeAssignmentBanner.swift
import SwiftUI

struct HomeAssignmentBanner: View {
    let baseID: String
    let baseName: String

    @Environment(AssignmentProfileStore.self) private var assignmentProfileStore

    private var profile: AssignmentProfile {
        assignmentProfileStore.profile(for: baseID)
    }

    var body: some View {
        if let message = bannerMessage {
            HStack(alignment: .top, spacing: 12) {
                Image(systemName: bannerIcon)
                    .font(.title3)
                    .foregroundStyle(AppTheme.accent)

                VStack(alignment: .leading, spacing: 4) {
                    Text(message.title)
                        .font(.subheadline.weight(.semibold))
                    Text(message.subtitle)
                        .font(.caption)
                        .foregroundStyle(.secondary)
                }

                Spacer(minLength: 0)
            }
            .padding(14)
            .appCardStyle(padding: 0, background: AppTheme.accent.opacity(0.08))
        }
    }

    private var bannerIcon: String {
        switch profile.phase {
        case .inbound: "airplane.arrival"
        case .stationed: "house.fill"
        case .outbound: "airplane.departure"
        }
    }

    private var bannerMessage: (title: String, subtitle: String)? {
        switch profile.phase {
        case .inbound:
            guard let reportDate = profile.reportDate else { return nil }
            return (
                countdownTitle(subject: "Report", date: reportDate),
                "In processing at \(baseName) · open Assignment for your checklist."
            )

        case .outbound:
            guard let pcsDate = profile.pcsDate else { return nil }
            return (
                countdownTitle(subject: "PCS", date: pcsDate),
                "Out processing from \(baseName) · use Leave Planner to check balances."
            )

        case .stationed:
            if let pcsDate = profile.pcsDate {
                let days = assignmentProfileStore.daysUntilPCS(for: baseID) ?? 0
                if days >= 0, days <= 120 {
                    return (
                        countdownTitle(subject: "PCS", date: pcsDate),
                        "Planning ahead · set phase to Out Processing when out-processing starts."
                    )
                }
            }
            return nil
        }
    }

    private func countdownTitle(subject: String, date: Date) -> String {
        let calendar = Calendar.current
        let days = calendar.dateComponents(
            [.day],
            from: calendar.startOfDay(for: Date()),
            to: calendar.startOfDay(for: date)
        ).day ?? 0

        switch days {
        case ..<0: return "\(subject) date has passed"
        case 0: return "\(subject) is today"
        case 1: return "1 day until \(subject)"
        default: return "\(days) days until \(subject)"
        }
    }
}

File: android/app/src/main/java/com/ryanladuca/myafbase/ui/home/HomeAssignmentBanner.kt
package com.ryanladuca.myafbase.ui.home

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.Assignment
import androidx.compose.material.icons.automirrored.filled.KeyboardArrowRight
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import com.ryanladuca.myafbase.data.db.AssignmentProfileEntity
import com.ryanladuca.myafbase.domain.model.AssignmentSegment
import java.util.Calendar
import java.util.concurrent.TimeUnit

@Composable
fun HomeAssignmentBanner(
    baseName: String,
    profile: AssignmentProfileEntity?,
    onOpenAssignment: () -> Unit,
) {
    val message = bannerMessage(baseName, profile) ?: return
    val scheme = MaterialTheme.colorScheme
    val phase = AssignmentSegment.fromRaw(profile?.phaseRaw)
    val containerColor = when (phase) {
        AssignmentSegment.INBOUND -> scheme.tertiaryContainer
        AssignmentSegment.STATIONED -> scheme.primaryContainer
        AssignmentSegment.OUTBOUND -> scheme.secondaryContainer
    }
    val contentColor = when (phase) {
        AssignmentSegment.INBOUND -> scheme.onTertiaryContainer
        AssignmentSegment.STATIONED -> scheme.onPrimaryContainer
        AssignmentSegment.OUTBOUND -> scheme.onSecondaryContainer
    }
    Surface(
        onClick = onOpenAssignment,
        modifier = Modifier.fillMaxWidth(),
        shape = MaterialTheme.shapes.extraLarge,
        color = containerColor,
        tonalElevation = 0.dp,
    ) {
        Row(
            modifier = Modifier
                .fillMaxWidth()
                .padding(16.dp),
            horizontalArrangement = Arrangement.spacedBy(12.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Surface(
                shape = MaterialTheme.shapes.medium,
                color = contentColor.copy(alpha = 0.12f),
                tonalElevation = 0.dp,
            ) {
                Icon(
                    imageVector = Icons.AutoMirrored.Filled.Assignment,
                    contentDescription = null,
                    tint = contentColor,
                    modifier = Modifier
                        .padding(10.dp)
                        .size(20.dp),
                )
            }
            Column(modifier = Modifier.weight(1f), verticalArrangement = Arrangement.spacedBy(4.dp)) {
                Text(
                    message.title,
                    style = MaterialTheme.typography.titleMedium,
                    fontWeight = FontWeight.SemiBold,
                    color = contentColor,
                )
                Text(
                    message.subtitle,
                    style = MaterialTheme.typography.bodySmall,
                    color = contentColor.copy(alpha = 0.82f),
                )
            }
            Icon(
                Icons.AutoMirrored.Filled.KeyboardArrowRight,
                contentDescription = null,
                tint = contentColor,
            )
        }
    }
}

private data class BannerMessage(val title: String, val subtitle: String)

private fun bannerMessage(baseName: String, profile: AssignmentProfileEntity?): BannerMessage? {
    val phase = AssignmentSegment.fromRaw(profile?.phaseRaw)
    return when (phase) {
        AssignmentSegment.INBOUND -> profile?.reportDateMillis?.let { date ->
            BannerMessage(
                countdownTitle("Report", date),
                "In processing at $baseName · open Assignment for your checklist.",
            )
        }
        AssignmentSegment.OUTBOUND -> profile?.pcsDateMillis?.let { date ->
            BannerMessage(
                countdownTitle("PCS", date),
                "Out processing from $baseName · use Leave Planner to check balances.",
            )
        }
        AssignmentSegment.STATIONED -> profile?.pcsDateMillis?.let { date ->
            val days = daysUntil(date)
            if (days in 0..120) {
                BannerMessage(
                    countdownTitle("PCS", date),
                    "Planning ahead · set phase to Out Processing when out-processing starts.",
                )
            } else null
        }
    }
}

private fun daysUntil(millis: Long): Int {
    val now = startOfDay(System.currentTimeMillis())
    val target = startOfDay(millis)
    return TimeUnit.MILLISECONDS.toDays(target - now).toInt()
}

private fun startOfDay(millis: Long): Long {
    val cal = Calendar.getInstance().apply { timeInMillis = millis }
    cal.set(Calendar.HOUR_OF_DAY, 0)
    cal.set(Calendar.MINUTE, 0)
    cal.set(Calendar.SECOND, 0)
    cal.set(Calendar.MILLISECOND, 0)
    return cal.timeInMillis
}

private fun countdownTitle(subject: String, dateMillis: Long): String {
    val days = daysUntil(dateMillis)
    return when {
        days < 0 -> "$subject was ${-days} days ago"
        days == 0 -> "$subject is today"
        days == 1 -> "$subject in 1 day"
        else -> "$subject in $days days"
    }
}
